keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -303,6 +318,29 @@ type Webhook struct {
 }
 ```
 
+### Convert to webhook-requested version
+
+Webhooks currently register to intercept particular API group/version/resource combinations.
+
+Some resources can be accessed via different versions, or even different API 
+groups (for example, `apps/v1` and `extensions/v1beta1` Deployments). To 
+intercept a resource effectively, all accessible groups/versions/resources
+must be registered for and understood by the webhook.
+
+When upgrading to a new version of the apiserver, existing resources can be 
+made available via new versions (or even new groups). Ensuring all webhooks
+(and registered webhook configurations) have been updated to be able to 
+handle the new versions/groups in every upgrade is possible, but easy to 
+forget to do, or to do incorrectly. In the case of webhooks not authored 
+by the cluster-administrator, obtaining updated admission plugins that 
+understand the new versions could require significant effort and time.
+
+Since the apiserver can convert between all of the versions by which a resource 
+is made available, this situation can be improved by having the apiserver 
+convert resources to the group/versions a webhook registered for.
+
+The API representation and behavior for this feature is still under design and will be updated/approved here prior to implementation.
+
 ## V1 API
 
 The currently planned `v1` API is described in this section.
```

Comment:
IMO, any api in the KEP is just a handwavy hint; you don't know the API for sure until you've gotten through the API review process. Personally I'd probably just list the features the API will address. Don't block the KEP on an API review; don't tie the API reviewer's hands with a KEP.

(This is more a general comment on KEPs than this specific one; obviously @liggitt probably has a good idea what is going to make it through the API review process. I just think this style of section is a poor example for other KEPs to copy.)

Reply:
In general, I'd rather see KEPs be *less* handwavy and actually put thought into the shape, behavior, and implications of a new API or an API change. They are a great way to get an early checkpoint from an API reviewer (xref the inline CSI volume API discussion in the KEP[[1]](https://github.com/kubernetes/enhancements/pull/716)[[2]](https://github.com/kubernetes/enhancements/pull/790), which prompted quite a bit of clarification and a couple significant changes prior to implementation).

I completely agree the API plan-of-record is just that... a plan... until the implementation/testing is complete (as are all aspects of KEPs that turn out to clash with reality), but as differences are encountered and worked through, bringing the current state/shape back to the KEP helps people following the feature who aren't reviewing every PR.